File: web/src/features/Search/__tests__/spies/spy-search-repository.ts
import type { AuthorityId, SearchResult } from '../../../../domain/types';
import type { SearchRepository } from '../../../../domain/ports/search-repository';

export class SpySearchRepository implements SearchRepository {
  searchCalls: Array<{ query: string; authorityId: AuthorityId; page: number }> = [];
  searchResult: SearchResult = { applications: [], total: 0, page: 1 };
  searchError: Error | null = null;

  async search(query: string, authorityId: AuthorityId, page: number): Promise<SearchResult> {
    this.searchCalls.push({ query, authorityId, page });
    if (this.searchError) {
      throw this.searchError;
    }
    return this.searchResult;
  }
}

File: mobile/ios/town-crier-tests/Sources/Spies/SpySearchRepository.swift
import TownCrierDomain

final class SpySearchRepository: SearchRepository, @unchecked Sendable {
  struct SearchCall: Equatable {
    let query: String
    let authorityId: Int
    let page: Int
  }

  private(set) var searchCalls: [SearchCall] = []
  var searchResult: Result<SearchResult, Error> = .success(
    SearchResult(applications: [], total: 0, page: 1)
  )

  func search(query: String, authorityId: Int, page: Int) async throws -> SearchResult {
    searchCalls.append(SearchCall(query: query, authorityId: authorityId, page: page))
    return try searchResult.get()
  }
}
